File: rust/src/client_recovery.rs
use crate::{DeliveryCursor, DeliveryCursorTransition, advance_delivery_cursor};

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct ClientRecoveryState {
    pub cursor: Option<DeliveryCursor>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ClientRecoveryEntry<Stream> {
    pub stream: Stream,
    pub cursor: Option<DeliveryCursor>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ClientRecoveryEvent<Stream> {
    Update {
        stream: Stream,
        cursor: Option<DeliveryCursor>,
    },
    Recovery {
        entries: Vec<ClientRecoveryEntry<Stream>>,
        latest_cursor: Option<DeliveryCursor>,
        snapshot_required: bool,
    },
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ClientInvalidation<Stream> {
    None,
    Streams(Vec<Stream>),
    All,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ClientRecoveryDecision<Stream> {
    pub state: ClientRecoveryState,
    pub invalidation: ClientInvalidation<Stream>,
    pub requires_snapshot: bool,
}

fn accept_cursor(
    current: Option<&DeliveryCursor>,
    incoming: &DeliveryCursor,
) -> (DeliveryCursor, bool) {
    match advance_delivery_cursor(current, incoming) {
        DeliveryCursorTransition::Initialized(cursor)
        | DeliveryCursorTransition::Advanced(cursor)
        | DeliveryCursorTransition::Stale(cursor) => (cursor, false),
        DeliveryCursorTransition::EpochChanged(cursor) => (cursor, true),
    }
}

/// Reduces transport-decoded deliveries into deterministic invalidation work.
#[must_use]
pub fn reduce_client_recovery<Stream>(
    state: &ClientRecoveryState,
    event: &ClientRecoveryEvent<Stream>,
) -> ClientRecoveryDecision<Stream>
where
    Stream: Clone + Eq,
{
    match event {
        ClientRecoveryEvent::Update { stream, cursor } => {
            let Some(cursor) = cursor else {
                return ClientRecoveryDecision {
                    state: state.clone(),
                    invalidation: ClientInvalidation::Streams(vec![stream.clone()]),
                    requires_snapshot: false,
                };
            };
            let (cursor, epoch_changed) = accept_cursor(state.cursor.as_ref(), cursor);
            ClientRecoveryDecision {
                state: ClientRecoveryState {
                    cursor: Some(cursor),
                },
                invalidation: if epoch_changed {
                    ClientInvalidation::All
                } else {
                    ClientInvalidation::Streams(vec![stream.clone()])
                },
                requires_snapshot: epoch_changed,
            }
        }
        ClientRecoveryEvent::Recovery {
            entries,
            latest_cursor,
            snapshot_required,
        } => {
            if *snapshot_required {
                return ClientRecoveryDecision {
                    state: ClientRecoveryState {
                        cursor: latest_cursor.clone(),
                    },
                    invalidation: ClientInvalidation::All,
                    requires_snapshot: true,
                };
            }

            let mut cursor = state.cursor.clone();
            let mut epoch_changed = false;
            let mut streams = Vec::new();
            for entry in entries {
                if !streams.contains(&entry.stream) {
                    streams.push(entry.stream.clone());
                }
                if let Some(incoming) = &entry.cursor {
                    let accepted = accept_cursor(cursor.as_ref(), incoming);
                    cursor = Some(accepted.0);
                    epoch_changed |= accepted.1;
                }
            }
            if let Some(incoming) = latest_cursor {
                let accepted = accept_cursor(cursor.as_ref(), incoming);
                cursor = Some(accepted.0);
                epoch_changed |= accepted.1;
            }

            ClientRecoveryDecision {
                state: ClientRecoveryState { cursor },
                invalidation: if epoch_changed {
                    ClientInvalidation::All
                } else if streams.is_empty() {
                    ClientInvalidation::None
                } else {
                    ClientInvalidation::Streams(streams)
                },
                requires_snapshot: epoch_changed,
            }
        }
    }
}

File: typescript/src/client-recovery.ts
import { advanceDeliveryCursor, type DeliveryCursor } from "./recovery.js";

export interface ClientRecoveryState {
  readonly cursor: DeliveryCursor | null;
}

export interface ClientRecoveryEntry<Stream> {
  readonly stream: Stream;
  readonly cursor?: DeliveryCursor;
}

export type ClientRecoveryEvent<Stream> =
  | {
      readonly kind: "update";
      readonly stream: Stream;
      readonly cursor?: DeliveryCursor;
    }
  | {
      readonly kind: "recovery";
      readonly entries: readonly ClientRecoveryEntry<Stream>[];
      readonly latestCursor?: DeliveryCursor;
      readonly snapshotRequired?: boolean;
    };

export type ClientInvalidation<Stream> =
  | { readonly kind: "none" }
  | { readonly kind: "streams"; readonly streams: readonly Stream[] }
  | { readonly kind: "all" };

export interface ClientRecoveryDecision<Stream> {
  readonly state: ClientRecoveryState;
  readonly invalidation: ClientInvalidation<Stream>;
  /** True when continuity was lost and consumers must read authoritative snapshots. */
  readonly requiresSnapshot: boolean;
}

export function createClientRecoveryState(
  cursor: DeliveryCursor | null = null,
): ClientRecoveryState {
  if (cursor === null) return Object.freeze({ cursor: null });
  const transition = advanceDeliveryCursor(null, cursor);
  return Object.freeze({ cursor: transition.cursor });
}

function acceptCursor(
  current: DeliveryCursor | null,
  incoming: DeliveryCursor,
): { readonly cursor: DeliveryCursor; readonly epochChanged: boolean } {
  const transition = advanceDeliveryCursor(current, incoming);
  return {
    cursor: transition.cursor,
    epochChanged: transition.kind === "epoch-changed",
  };
}

/** Reduces transport-decoded delivery events into deterministic invalidation work. */
export function reduceClientRecovery<Stream>(
  state: ClientRecoveryState,
  event: ClientRecoveryEvent<Stream>,
): ClientRecoveryDecision<Stream> {
  if (event.kind === "update") {
    if (event.cursor === undefined) {
      return {
        state,
        invalidation: { kind: "streams", streams: [event.stream] },
        requiresSnapshot: false,
      };
    }
    const accepted = acceptCursor(state.cursor, event.cursor);
    return {
      state: createClientRecoveryState(accepted.cursor),
      invalidation: accepted.epochChanged
        ? { kind: "all" }
        : { kind: "streams", streams: [event.stream] },
      requiresSnapshot: accepted.epochChanged,
    };
  }

  if (event.snapshotRequired === true) {
    return {
      state: createClientRecoveryState(event.latestCursor ?? null),
      invalidation: { kind: "all" },
      requiresSnapshot: true,
    };
  }

  let cursor = state.cursor;
  let epochChanged = false;
  const streams = new Set<Stream>();
  for (const entry of event.entries) {
    streams.add(entry.stream);
    if (entry.cursor !== undefined) {
      const accepted = acceptCursor(cursor, entry.cursor);
      cursor = accepted.cursor;
      epochChanged ||= accepted.epochChanged;
    }
  }
  if (event.latestCursor !== undefined) {
    const accepted = acceptCursor(cursor, event.latestCursor);
    cursor = accepted.cursor;
    epochChanged ||= accepted.epochChanged;
  }

  return {
    state: createClientRecoveryState(cursor),
    invalidation: epochChanged
      ? { kind: "all" }
      : streams.size === 0
        ? { kind: "none" }
        : { kind: "streams", streams: [...streams] },
    requiresSnapshot: epochChanged,
  };
}
